fgpyo/util/metric.py

```diff
@@ -334,3 +346,55 @@ def fast_concat(*inputs: Path, output: Path) -> None:
             io.write_lines(
                 path=output, lines_to_write=list(io.read_lines(input_path))[1:], append=True
             )
+
+
+def is_dataclass_instance(metric: Metric) -> TypeGuard[DataclassInstance]:
+    """
+    Test if the given metric is a dataclass instance.
+
+    NB: `dataclasses.is_dataclass` returns True for both dataclass instances and class objects, and
+    we need to override the built-in function's `TypeGuard`.
+
+    Args:
+        metric: An instance of a Metric.
+
+    Returns:
+        True if the given metric is an instance of a dataclass-decorated Metric.
+        False otherwise.
+    """
+    return not isclass(metric) and dataclasses.is_dataclass(metric)
+
+
+def is_attrs_instance(metric: Metric) -> TypeGuard[AttrsInstance]:
+    """
+    Test if the given metric is an attr.s instance.
+
+    NB: `attr.has` provides a type guard, but only on the class object - we want to narrow the type
+    of the metric instance, so we implement a guard here.
+
+    Args:
+        metric: An instance of a Metric.
+
+    Returns:
+        True if the given metric is an instance of an attr.s-decorated Metric.
+        False otherwise.
+    """
+    return not isclass(metric) and attr.has(metric.__class__)
+
+
+def asdict(metric: Metric) -> Dict[str, Any]:
+    """
+    Convert a Metric instance to a dictionary.
+
+    Args:
+        metric: An instance of a Metric.
+
+    Returns:
+        A dictionary representation of the given metric.
+    """
```

Comment:
This overlooks the `format_value` method on `Metric` that is used to format values when written to a file.

Reply:
@nh13 the omission was deliberate. I left a comment on the topic in the PR description, but given how much conversation this one has attracted it was easy to overlook 🙂 

> NB: it may be valuable to permit formatting/casting the values as part of this function (e.g. by adding a parameter `format: bool = False`), but to do so we'll probably want to extract `Metric.format_value()` to a standalone function instead of a classmethod

I do not think we should have an `asdict()` function that changes the value types by default. This function is primarily intended as a dispatcher, selecting the correct (`dataclasses` or `attr.s`) function depending on how the `Metric` in question was selected. 

Ideally, this function will be deprecated once we drop support for `attrs`. As I mentioned in my comments above to Clint, when that happens it would be preferable to be able to transparently replace this with `dataclasses.asdict` 

> At that time we could simply import `dataclasses.asdict` into this module to avoid breakage? Or replace `from fgpyo.util.metric import asdict` with `from dataclasses import asdict`

I am open to adding `format: bool = False` to this function, with the stipulation that it should be `False` by default and the caveat that I expect it to add debt and increase friction when we deprecate `attr.s`.

However, I'd prefer to leave as is, and then call `format_value` on the resulting dict when necessary, e.g.
```py
metric_dict = {key: format_value(val) for key, val in asdict(metric)}
```
as I find the above relatively easy to understand

Thoughts?

Reply:
Let’s omit it but document it clearly. 

I do like that the format_value is a class method so that any new Metric type can perform custom formatting but overriding the class method. I think passing in a parsing function, like defopt, in rare situations causes a conflict when we want to format the type differently.

Reply:
I like the idea of overriding the classmethod, but can't imagine doing so in practice. Given that it specifies formatting behavior for a wide variety of primitive and compound types, it doesn't seem to lend itself to easy extension or modification. (Since in order to override formatting for one type, you would have to override them all.)